# Lab 5: Exploring the Stack Overflow Developer Survey Dataset
## IBM Data Analytics Capstone Project
### Objective: Load, explore and summarize the Stack Overflow Developer Survey dataset


In [1]:
import pandas as pd
print("Pandas version:", pd.__version__)

Pandas version: 2.3.2


In [2]:
# Load the dataset directly from the URL
file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/VYPrOu0Vs3I0hKLLjiPGrA/survey-data-with-duplicate.csv"

df = pd.read_csv(file_path)
print(" Dataset loaded successfully!")

 Dataset loaded successfully!


In [3]:
# Display the top 5 rows of the dataset
df.head()

,ResponseId,MainBranch,Age,Employment,RemoteWork,Check,CodingActivities,EdLevel,LearnCode,LearnCodeOnline,...,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,SurveyLength,SurveyEase,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,Under 18 years old,"Employed, full-time",Remote,Apples,Hobby,Primary/elementary school,Books / Physical media,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,I am a developer by profession,35-44 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
2,3,I am a developer by profession,45-54 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,NaN,NaN,NaN,NaN,NaN,NaN,Appropriate in length,Easy,NaN,NaN
3,4,I am learning to code,18-24 years old,"Student, full-time",NaN,Apples,NaN,Some college/university study without earning ...,"Other online resources (e.g., videos, blogs, f...",Stack Overflow;How-to videos;Interactive tutorial,...,NaN,NaN,NaN,NaN,NaN,NaN,Too long,Easy,NaN,NaN
4,5,I am a developer by profession,18-24 years old,"Student, full-time",NaN,Apples,NaN,"Secondary school (e.g. American high school, G...","Other online resources (e.g., videos, blogs, f...",Technical documentation;Blogs;Written Tutorial...,...,NaN,NaN,NaN,NaN,NaN,NaN,Too short,Easy,NaN,NaN


In [4]:
# Print the number of rows and columns
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 65457
Number of columns: 114


In [5]:
# Print the datatype of all columns
print(df.dtypes)

ResponseId               int64
MainBranch              object
Age                     object
Employment              object
RemoteWork              object
                        ...   
JobSatPoints_11        float64
SurveyLength            object
SurveyEase              object
ConvertedCompYearly    float64
JobSat                 float64
Length: 114, dtype: object


In [6]:
# Print the mean age of the survey participants
print("Mean age of survey participants:")
print(df['Age'].value_counts())
print("\nNote: Age is stored as a range (object type), not a numeric value")

Mean age of survey participants:
Age
25-34 years old       23913
35-44 years old       14948
18-24 years old       14105
45-54 years old        6251
55-64 years old        2575
Under 18 years old     2570
65 years or older       773
Prefer not to say       322
Name: count, dtype: int64

Note: Age is stored as a range (object type), not a numeric value


In [7]:
# Print the number of unique countries
print("Number of unique countries:", df['Country'].nunique())
print("\nTop 10 countries by responses:")
print(df['Country'].value_counts().head(10))

Number of unique countries: 185

Top 10 countries by responses:
Country
United States of America                                11099
Germany                                                  4949
India                                                    4232
United Kingdom of Great Britain and Northern Ireland     3227
Ukraine                                                  2672
France                                                   2110
Canada                                                   2106
Poland                                                   1534
Netherlands                                              1449
Brazil                                                   1375
Name: count, dtype: int64


## Key Findings from Dataset Exploration

- **Total Responses:** 65,457 developers worldwide
- **Total Columns:** 114 features
- **Countries Represented:** 185 unique countries
- **Top Country:** USA with 11,099 responses
- **Largest Age Group:** 25-34 years old (23,913 respondents)
- **Data Types:** Mix of int64, float64, and object columns
- **Dataset Source:** Stack Overflow Developer Survey (ODbL License)